{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [],
   "source": [
    "'''\n",
    "First define a function to read in fasta file.\n",
    "This is very similar to your code, with two major differences:\n",
    "1. adding line = line.rstrip(), which will remove \\n at the end of each line\n",
    "2. save results as a generator object, with name (strains) and seq (genomes), easier to store and reference\n",
    "\n",
    "'''\n",
    "import pandas as pd\n",
    "def read_fasta(fp):\n",
    "        name, seq = None, []\n",
    "        for line in fp:\n",
    "            line = line.rstrip()\n",
    "            if line.startswith(\">\"):\n",
    "                if name: yield (name, ''.join(seq))\n",
    "                name, seq = line, []\n",
    "            else:\n",
    "                seq.append(line)\n",
    "        if name: yield (name, ''.join(seq))\n",
    "\n",
    "            "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Extra: \n",
    "BioPython is a library designed for reading in fasta format data faster, and easier since it has methods wrapped up. Here is a tutorial if would like to explore more: https://biopython.org/DIST/docs/tutorial/Tutorial.html#htoc11"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "nucleotides appeared in raw data:\n",
      "{'A': '1', 'G': '2', 'C': '3', 'T': '4', 'N': '5', 'S': '5', 'R': '5', 'M': '5', 'Y': '5', 'W': '5', 'K': '5', 'B': '5', 'D': '5'}\n"
     ]
    }
   ],
   "source": [
    "'''\n",
    "After reading in data, we could observe data quality.\n",
    "\n",
    "We would find the unique nucleotides in raw data is more than just \"A/G/C/T/N\". \n",
    "Here we assign a string \"5\" to all other nucleotides. \n",
    "We store nucleotides character and numerical mapping into a dictionary for later use:\n",
    "key is the character such as \"A\"/\"G\",\n",
    "value is the numerical mapping such as \"1\"/\"2\".\n",
    "\n",
    "'''\n",
    "\n",
    "strains = []\n",
    "genomes = []\n",
    "nucleotides = {\"A\": \"1\", \"G\": \"2\", \"C\": \"3\", \"T\": \"4\"}\n",
    "with open(r'C:/Users/srava/BedfordProjects/zika/results/aligned.fasta') as fp:\n",
    "    for name, seq in read_fasta(fp):\n",
    "        strains.append(name)\n",
    "        this_nu = list(set(seq))\n",
    "        for nu in this_nu:\n",
    "            if nu not in nucleotides: #assign \"5\" to all other nucleotides\n",
    "                nucleotides[nu] = \"5\"\n",
    "                \n",
    "print(\"nucleotides appeared in raw data:\")\n",
    "print(nucleotides)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[5. 5. 5. ... 5. 5. 5.]\n"
     ]
    }
   ],
   "source": [
    "import numpy as np\n",
    "genomes = []\n",
    "0\n",
    "with open(r'C:/Users/srava/BedfordProjects/zika/results/aligned.fasta') as fp:\n",
    "    for name, seq in read_fasta(fp):        \n",
    "        for nu in nucleotides:\n",
    "            seq = seq.replace(\"A\",\"1\").replace('G','2').replace('C','3').replace('T','4').replace('N','5').replace('R','5').replace('S','5').replace('M','5').replace('Y','5').replace('W','5').replace('K', '5').replace('B','5').replace('D','5')\n",
    "            genomes.append(list(seq))\n",
    "genomes = np.array(genomes, dtype = np.float64)\n",
    "print(genomes[1])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "      0      1      2      3      4      5      6      7      8      9      \\\n",
      "0       5.0    5.0    5.0    5.0    5.0    5.0    5.0    5.0    5.0    5.0   \n",
      "1       5.0    5.0    5.0    5.0    5.0    5.0    5.0    5.0    5.0    5.0   \n",
      "2       5.0    5.0    5.0    5.0    5.0    5.0    5.0    5.0    5.0    5.0   \n",
      "3       5.0    5.0    5.0    5.0    5.0    5.0    5.0    5.0    5.0    5.0   \n",
      "4       5.0    5.0    5.0    5.0    5.0    5.0    5.0    5.0    5.0    5.0   \n",
      "5       5.0    5.0    5.0    5.0    5.0    5.0    5.0    5.0    5.0    5.0   \n",
      "6       5.0    5.0    5.0    5.0    5.0    5.0    5.0    5.0    5.0    5.0   \n",
      "7       5.0    5.0    5.0    5.0    5.0    5.0    5.0    5.0    5.0    5.0   \n",
      "8       5.0    5.0    5.0    5.0    5.0    5.0    5.0    5.0    5.0    5.0   \n",
      "9       5.0    5.0    5.0    5.0    5.0    5.0    5.0    5.0    5.0    5.0   \n",
      "10      5.0    5.0    5.0    5.0    5.0    5.0    5.0    5.0    5.0    5.0   \n",
      "11      5.0    5.0    5.0    5.0    5.0    5.0    5.0    5.0    5.0    5.0   \n",
      "12      5.0    5.0    5.0    5.0    5.0    5.0    5.0    5.0    5.0    5.0   \n",
      "13      2.0    1.0    1.0    4.0    3.0    1.0    2.0    1.0    3.0    4.0   \n",
      "14      2.0    1.0    1.0    4.0    3.0    1.0    2.0    1.0    3.0    4.0   \n",
      "15      2.0    1.0    1.0    4.0    3.0    1.0    2.0    1.0    3.0    4.0   \n",
      "16      2.0    1.0    1.0    4.0    3.0    1.0    2.0    1.0    3.0    4.0   \n",
      "17      2.0    1.0    1.0    4.0    3.0    1.0    2.0    1.0    3.0    4.0   \n",
      "18      2.0    1.0    1.0    4.0    3.0    1.0    2.0    1.0    3.0    4.0   \n",
      "19      2.0    1.0    1.0    4.0    3.0    1.0    2.0    1.0    3.0    4.0   \n",
      "20      2.0    1.0    1.0    4.0    3.0    1.0    2.0    1.0    3.0    4.0   \n",
      "21      2.0    1.0    1.0    4.0    3.0    1.0    2.0    1.0    3.0    4.0   \n",
      "22      2.0    1.0    1.0    4.0    3.0    1.0    2.0    1.0    3.0    4.0   \n",
      "23      2.0    1.0    1.0    4.0    3.0    1.0    2.0    1.0    3.0    4.0   \n",
      "24      2.0    1.0    1.0    4.0    3.0    1.0    2.0    1.0    3.0    4.0   \n",
      "25      2.0    1.0    1.0    4.0    3.0    1.0    2.0    1.0    3.0    4.0   \n",
      "26      2.0    1.0    1.0    4.0    3.0    1.0    2.0    1.0    3.0    4.0   \n",
      "27      2.0    1.0    1.0    4.0    3.0    1.0    2.0    1.0    3.0    4.0   \n",
      "28      2.0    1.0    1.0    4.0    3.0    1.0    2.0    1.0    3.0    4.0   \n",
      "29      2.0    1.0    1.0    4.0    3.0    1.0    2.0    1.0    3.0    4.0   \n",
      "...     ...    ...    ...    ...    ...    ...    ...    ...    ...    ...   \n",
      "6717    5.0    5.0    5.0    5.0    5.0    5.0    5.0    5.0    5.0    5.0   \n",
      "6718    5.0    5.0    5.0    5.0    5.0    5.0    5.0    5.0    5.0    5.0   \n",
      "6719    5.0    5.0    5.0    5.0    5.0    5.0    5.0    5.0    5.0    5.0   \n",
      "6720    5.0    5.0    5.0    5.0    5.0    5.0    5.0    5.0    5.0    5.0   \n",
      "6721    5.0    5.0    5.0    5.0    5.0    5.0    5.0    5.0    5.0    5.0   \n",
      "6722    5.0    5.0    5.0    5.0    5.0    5.0    5.0    5.0    5.0    5.0   \n",
      "6723    5.0    5.0    5.0    5.0    5.0    5.0    5.0    5.0    5.0    5.0   \n",
      "6724    5.0    5.0    5.0    5.0    5.0    5.0    5.0    5.0    5.0    5.0   \n",
      "6725    5.0    5.0    5.0    5.0    5.0    5.0    5.0    5.0    5.0    5.0   \n",
      "6726    5.0    5.0    5.0    5.0    5.0    5.0    5.0    5.0    5.0    5.0   \n",
      "6727    5.0    5.0    5.0    5.0    5.0    5.0    5.0    5.0    5.0    5.0   \n",
      "6728    5.0    5.0    5.0    5.0    5.0    5.0    5.0    5.0    5.0    5.0   \n",
      "6729    5.0    5.0    5.0    5.0    5.0    5.0    5.0    5.0    5.0    5.0   \n",
      "6730    5.0    5.0    5.0    5.0    5.0    5.0    5.0    5.0    5.0    5.0   \n",
      "6731    5.0    5.0    5.0    5.0    5.0    5.0    5.0    5.0    5.0    5.0   \n",
      "6732    5.0    5.0    5.0    5.0    5.0    5.0    5.0    5.0    5.0    5.0   \n",
      "6733    5.0    5.0    5.0    5.0    5.0    5.0    5.0    5.0    5.0    5.0   \n",
      "6734    5.0    5.0    5.0    5.0    5.0    5.0    5.0    5.0    5.0    5.0   \n",
      "6735    5.0    5.0    5.0    5.0    5.0    5.0    5.0    5.0    5.0    5.0   \n",
      "6736    5.0    5.0    5.0    5.0    5.0    5.0    5.0    5.0    5.0    5.0   \n",
      "6737    5.0    5.0    5.0    5.0    5.0    5.0    5.0    5.0    5.0    5.0   \n",
      "6738    5.0    5.0    5.0    5.0    5.0    5.0    5.0    5.0    5.0    5.0   \n",
      "6739    5.0    5.0    5.0    5.0    5.0    5.0    5.0    5.0    5.0    5.0   \n",
      "6740    5.0    5.0    5.0    5.0    5.0    5.0    5.0    5.0    5.0    5.0   \n",
      "6741    5.0    5.0    5.0    5.0    5.0    5.0    5.0    5.0    5.0    5.0   \n",
      "6742    5.0    5.0    5.0    5.0    5.0    5.0    5.0    5.0    5.0    5.0   \n",
      "6743    5.0    5.0    5.0    5.0    5.0    5.0    5.0    5.0    5.0    5.0   \n",
      "6744    5.0    5.0    5.0    5.0    5.0    5.0    5.0    5.0    5.0    5.0   \n",
      "6745    5.0    5.0    5.0    5.0    5.0    5.0    5.0    5.0    5.0    5.0   \n",
      "6746    5.0    5.0    5.0    5.0    5.0    5.0    5.0    5.0    5.0    5.0   \n",
      "\n",
      "      ...  10759  10760  10761  10762  10763  10764  10765  10766  10767  \\\n",
      "0     ...    5.0    5.0    5.0    5.0    5.0    5.0    5.0    5.0    5.0   \n",
      "1     ...    5.0    5.0    5.0    5.0    5.0    5.0    5.0    5.0    5.0   \n",
      "2     ...    5.0    5.0    5.0    5.0    5.0    5.0    5.0    5.0    5.0   \n",
      "3     ...    5.0    5.0    5.0    5.0    5.0    5.0    5.0    5.0    5.0   \n",
      "4     ...    5.0    5.0    5.0    5.0    5.0    5.0    5.0    5.0    5.0   \n",
      "5     ...    5.0    5.0    5.0    5.0    5.0    5.0    5.0    5.0    5.0   \n",
      "6     ...    5.0    5.0    5.0    5.0    5.0    5.0    5.0    5.0    5.0   \n",
      "7     ...    5.0    5.0    5.0    5.0    5.0    5.0    5.0    5.0    5.0   \n",
      "8     ...    5.0    5.0    5.0    5.0    5.0    5.0    5.0    5.0    5.0   \n",
      "9     ...    5.0    5.0    5.0    5.0    5.0    5.0    5.0    5.0    5.0   \n",
      "10    ...    5.0    5.0    5.0    5.0    5.0    5.0    5.0    5.0    5.0   \n",
      "11    ...    5.0    5.0    5.0    5.0    5.0    5.0    5.0    5.0    5.0   \n",
      "12    ...    5.0    5.0    5.0    5.0    5.0    5.0    5.0    5.0    5.0   \n",
      "13    ...    3.0    2.0    2.0    3.0    2.0    2.0    3.0    3.0    2.0   \n",
      "14    ...    3.0    2.0    2.0    3.0    2.0    2.0    3.0    3.0    2.0   \n",
      "15    ...    3.0    2.0    2.0    3.0    2.0    2.0    3.0    3.0    2.0   \n",
      "16    ...    3.0    2.0    2.0    3.0    2.0    2.0    3.0    3.0    2.0   \n",
      "17    ...    3.0    2.0    2.0    3.0    2.0    2.0    3.0    3.0    2.0   \n",
      "18    ...    3.0    2.0    2.0    3.0    2.0    2.0    3.0    3.0    2.0   \n",
      "19    ...    3.0    2.0    2.0    3.0    2.0    2.0    3.0    3.0    2.0   \n",
      "20    ...    3.0    2.0    2.0    3.0    2.0    2.0    3.0    3.0    2.0   \n",
      "21    ...    3.0    2.0    2.0    3.0    2.0    2.0    3.0    3.0    2.0   \n",
      "22    ...    3.0    2.0    2.0    3.0    2.0    2.0    3.0    3.0    2.0   \n",
      "23    ...    3.0    2.0    2.0    3.0    2.0    2.0    3.0    3.0    2.0   \n",
      "24    ...    3.0    2.0    2.0    3.0    2.0    2.0    3.0    3.0    2.0   \n",
      "25    ...    3.0    2.0    2.0    3.0    2.0    2.0    3.0    3.0    2.0   \n",
      "26    ...    3.0    2.0    2.0    3.0    2.0    2.0    3.0    3.0    2.0   \n",
      "27    ...    3.0    2.0    2.0    3.0    2.0    2.0    3.0    3.0    2.0   \n",
      "28    ...    3.0    2.0    2.0    3.0    2.0    2.0    3.0    3.0    2.0   \n",
      "29    ...    3.0    2.0    2.0    3.0    2.0    2.0    3.0    3.0    2.0   \n",
      "...   ...    ...    ...    ...    ...    ...    ...    ...    ...    ...   \n",
      "6717  ...    5.0    5.0    5.0    5.0    5.0    5.0    5.0    5.0    5.0   \n",
      "6718  ...    5.0    5.0    5.0    5.0    5.0    5.0    5.0    5.0    5.0   \n",
      "6719  ...    5.0    5.0    5.0    5.0    5.0    5.0    5.0    5.0    5.0   \n",
      "6720  ...    5.0    5.0    5.0    5.0    5.0    5.0    5.0    5.0    5.0   \n",
      "6721  ...    5.0    5.0    5.0    5.0    5.0    5.0    5.0    5.0    5.0   \n",
      "6722  ...    5.0    5.0    5.0    5.0    5.0    5.0    5.0    5.0    5.0   \n",
      "6723  ...    5.0    5.0    5.0    5.0    5.0    5.0    5.0    5.0    5.0   \n",
      "6724  ...    5.0    5.0    5.0    5.0    5.0    5.0    5.0    5.0    5.0   \n",
      "6725  ...    5.0    5.0    5.0    5.0    5.0    5.0    5.0    5.0    5.0   \n",
      "6726  ...    5.0    5.0    5.0    5.0    5.0    5.0    5.0    5.0    5.0   \n",
      "6727  ...    5.0    5.0    5.0    5.0    5.0    5.0    5.0    5.0    5.0   \n",
      "6728  ...    5.0    5.0    5.0    5.0    5.0    5.0    5.0    5.0    5.0   \n",
      "6729  ...    5.0    5.0    5.0    5.0    5.0    5.0    5.0    5.0    5.0   \n",
      "6730  ...    5.0    5.0    5.0    5.0    5.0    5.0    5.0    5.0    5.0   \n",
      "6731  ...    5.0    5.0    5.0    5.0    5.0    5.0    5.0    5.0    5.0   \n",
      "6732  ...    5.0    5.0    5.0    5.0    5.0    5.0    5.0    5.0    5.0   \n",
      "6733  ...    5.0    5.0    5.0    5.0    5.0    5.0    5.0    5.0    5.0   \n",
      "6734  ...    5.0    5.0    5.0    5.0    5.0    5.0    5.0    5.0    5.0   \n",
      "6735  ...    5.0    5.0    5.0    5.0    5.0    5.0    5.0    5.0    5.0   \n",
      "6736  ...    5.0    5.0    5.0    5.0    5.0    5.0    5.0    5.0    5.0   \n",
      "6737  ...    5.0    5.0    5.0    5.0    5.0    5.0    5.0    5.0    5.0   \n",
      "6738  ...    5.0    5.0    5.0    5.0    5.0    5.0    5.0    5.0    5.0   \n",
      "6739  ...    5.0    5.0    5.0    5.0    5.0    5.0    5.0    5.0    5.0   \n",
      "6740  ...    5.0    5.0    5.0    5.0    5.0    5.0    5.0    5.0    5.0   \n",
      "6741  ...    5.0    5.0    5.0    5.0    5.0    5.0    5.0    5.0    5.0   \n",
      "6742  ...    5.0    5.0    5.0    5.0    5.0    5.0    5.0    5.0    5.0   \n",
      "6743  ...    5.0    5.0    5.0    5.0    5.0    5.0    5.0    5.0    5.0   \n",
      "6744  ...    5.0    5.0    5.0    5.0    5.0    5.0    5.0    5.0    5.0   \n",
      "6745  ...    5.0    5.0    5.0    5.0    5.0    5.0    5.0    5.0    5.0   \n",
      "6746  ...    5.0    5.0    5.0    5.0    5.0    5.0    5.0    5.0    5.0   \n",
      "\n",
      "      10768  \n",
      "0       5.0  \n",
      "1       5.0  \n",
      "2       5.0  \n",
      "3       5.0  \n",
      "4       5.0  \n",
      "5       5.0  \n",
      "6       5.0  \n",
      "7       5.0  \n",
      "8       5.0  \n",
      "9       5.0  \n",
      "10      5.0  \n",
      "11      5.0  \n",
      "12      5.0  \n",
      "13      2.0  \n",
      "14      2.0  \n",
      "15      2.0  \n",
      "16      2.0  \n",
      "17      2.0  \n",
      "18      2.0  \n",
      "19      2.0  \n",
      "20      2.0  \n",
      "21      2.0  \n",
      "22      2.0  \n",
      "23      2.0  \n",
      "24      2.0  \n",
      "25      2.0  \n",
      "26      2.0  \n",
      "27      2.0  \n",
      "28      2.0  \n",
      "29      2.0  \n",
      "...     ...  \n",
      "6717    5.0  \n",
      "6718    5.0  \n",
      "6719    5.0  \n",
      "6720    5.0  \n",
      "6721    5.0  \n",
      "6722    5.0  \n",
      "6723    5.0  \n",
      "6724    5.0  \n",
      "6725    5.0  \n",
      "6726    5.0  \n",
      "6727    5.0  \n",
      "6728    5.0  \n",
      "6729    5.0  \n",
      "6730    5.0  \n",
      "6731    5.0  \n",
      "6732    5.0  \n",
      "6733    5.0  \n",
      "6734    5.0  \n",
      "6735    5.0  \n",
      "6736    5.0  \n",
      "6737    5.0  \n",
      "6738    5.0  \n",
      "6739    5.0  \n",
      "6740    5.0  \n",
      "6741    5.0  \n",
      "6742    5.0  \n",
      "6743    5.0  \n",
      "6744    5.0  \n",
      "6745    5.0  \n",
      "6746    5.0  \n",
      "\n",
      "[6747 rows x 10769 columns]\n"
     ]
    }
   ],
   "source": [
    "genomesDf = pd.DataFrame(genomes)\n",
    "print(genomesDf)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[5. 5. 5. ... 5. 5. 5.]\n",
      "(array([   27,    28,    32, ..., 10615, 10617, 10620], dtype=int64),)\n"
     ]
    }
   ],
   "source": [
    "#numpy vectorize it?\n",
    "#how do you get the number of A's in a column\n",
    "countA = 0\n",
    "countG = 0\n",
    "countC = 0\n",
    "countT = 0\n",
    "countN = 0\n",
    "countOther = 0\n",
    "StoreCounts = []\n",
    "StoreCountsArr = []\n",
    "for i in range(0,len(genomesSplitArr)):\n",
    "    print(df[[\"Site \" + str(j) for j in range(0,len(genomes))]])\n",
    "    print(genomes[i])\n",
    "    countA = (np.where(genomes[i]< 1.1))\n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
