
<!DOCTYPE html>

<html>
  <head>
    <meta charset="utf-8" />
    <meta name="viewport" content="width=device-width, initial-scale=1.0" />
    <title>src.embed module &#8212; EmbedGen  documentation</title>
    
  <link href="static/css/theme.css" rel="stylesheet" />
  <link href="static/css/index.c5995385ac14fb8791e8eb36b4908be2.css" rel="stylesheet" />

    
  <link rel="stylesheet"
    href="static/vendor/fontawesome/5.13.0/css/all.min.css">
  <link rel="preload" as="font" type="font/woff2" crossorigin
    href="static/vendor/fontawesome/5.13.0/webfonts/fa-solid-900.woff2">
  <link rel="preload" as="font" type="font/woff2" crossorigin
    href="static/vendor/fontawesome/5.13.0/webfonts/fa-brands-400.woff2">

    
      

    
    <link rel="stylesheet" href="static/pygments.css" type="text/css" />
    <link rel="stylesheet" href="static/sphinx-book-theme.e8f53015daec13862f6db5e763c41738.css" type="text/css" />
    
  <link rel="preload" as="script" href="static/js/index.1c5a1a01449ed65a7b51.js">

    <script id="documentation_options" data-url_root="./" src="static/documentation_options.js"></script>
    <script src="static/jquery.js"></script>
    <script src="static/underscore.js"></script>
    <script src="static/doctools.js"></script>
    <script src="static/sphinx-book-theme.12a9622fbb08dcb3a2a40b2c02b83a57.js"></script>
    <link rel="index" title="Index" href="genindex.html" />
    <link rel="search" title="Search" href="search.html" />
    <meta name="viewport" content="width=device-width, initial-scale=1" />
    <meta name="docsearch:language" content="en" />
    
  </head>
  <body data-spy="scroll" data-target="#bd-toc-nav" data-offset="80">
    
    <div class="container-fluid" id="banner"></div>

    

    <div class="container-xl">
      <div class="row">
          
<div class="col-12 col-md-3 bd-sidebar site-navigation show" id="site-navigation">
    
        <div class="navbar-brand-box">
    <a class="navbar-brand text-wrap" href="#">
      
      
      <h1 class="site-logo" id="site-title">EmbedGen  documentation</h1>
      
    </a>
</div><form class="bd-search d-flex align-items-center" action="search.html" method="get">
  <i class="icon fas fa-search"></i>
  <input type="search" class="form-control" name="q" id="search-input" placeholder="Search the docs ..." aria-label="Search the docs ..." autocomplete="off" >
</form><nav class="bd-links" id="bd-docs-nav" aria-label="Main">
    <div class="bd-toc-item active">
        
    </div>
</nav> <!-- To handle the deprecated key -->

<div class="navbar_extra_footer">
  Theme by the <a href="https://ebp.jupyterbook.org">Executable Book Project</a>
</div>

</div>


          


          
<main class="col py-md-3 pl-md-4 bd-content overflow-auto" role="main">
    
    <div class="topbar container-xl fixed-top">
    <div class="topbar-contents row">
        <div class="col-12 col-md-3 bd-topbar-whitespace site-navigation show"></div>
        <div class="col pl-md-4 topbar-main">
            
            <button id="navbar-toggler" class="navbar-toggler ml-0" type="button" data-toggle="collapse"
                data-toggle="tooltip" data-placement="bottom" data-target=".site-navigation" aria-controls="navbar-menu"
                aria-expanded="true" aria-label="Toggle navigation" aria-controls="site-navigation"
                title="Toggle navigation" data-toggle="tooltip" data-placement="left">
                <i class="fas fa-bars"></i>
                <i class="fas fa-arrow-left"></i>
                <i class="fas fa-arrow-up"></i>
            </button>
            
            
<div class="dropdown-buttons-trigger">
    <button id="dropdown-buttons-trigger" class="btn btn-secondary topbarbtn" aria-label="Download this page"><i
            class="fas fa-download"></i></button>

    <div class="dropdown-buttons">
        <!-- ipynb file if we had a myst markdown file -->
        
        <!-- Download raw file -->
        <a class="dropdown-buttons" href="_sources/index.rst.txt"><button type="button"
                class="btn btn-secondary topbarbtn" title="Download source file" data-toggle="tooltip"
                data-placement="left">.rst</button></a>
        <!-- Download PDF via print -->
        <button type="button" id="download-print" class="btn btn-secondary topbarbtn" title="Print to PDF"
            onClick="window.print()" data-toggle="tooltip" data-placement="left">.pdf</button>
    </div>
</div>

            <!-- Source interaction buttons -->

            <!-- Full screen (wrap in <a> to have style consistency -->

<a class="full-screen-button"><button type="button" class="btn btn-secondary topbarbtn" data-toggle="tooltip"
        data-placement="bottom" onclick="toggleFullScreen()" aria-label="Fullscreen mode"
        title="Fullscreen mode"><i
            class="fas fa-expand"></i></button></a>

            <!-- Launch buttons -->

        </div>

        <!-- Table of contents -->
        <div class="d-none d-md-block col-md-2 bd-toc show">
            
            <div class="tocsection onthispage pt-5 pb-3">
                <i class="fas fa-list"></i> Contents
            </div>
            <nav id="bd-toc-nav" aria-label="Page">
                <ul class="visible nav section-nav flex-column">
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#command-line-interface">
   Command line interface
  </a>
  <ul class="nav section-nav flex-column">
   <li class="toc-h3 nav-item toc-entry">
    <a class="reference internal nav-link" href="#Positional Arguments">
     Positional Arguments
    </a>
   </li>
   <li class="toc-h3 nav-item toc-entry">
    <a class="reference internal nav-link" href="#Named Arguments">
     Named Arguments
    </a>
   </li>
   <li class="toc-h3 nav-item toc-entry">
    <a class="reference internal nav-link" href="#Sub-commands:">
     Sub-commands:
    </a>
    <ul class="nav section-nav flex-column">
     <li class="toc-h4 nav-item toc-entry">
      <a class="reference internal nav-link" href="#pca">
       pca
      </a>
      <ul class="nav section-nav flex-column">
       <li class="toc-h5 nav-item toc-entry">
        <a class="reference internal nav-link" href="#Named Arguments_repeat1">
         Named Arguments
        </a>
       </li>
      </ul>
     </li>
     <li class="toc-h4 nav-item toc-entry">
      <a class="reference internal nav-link" href="#t-sne">
       t-sne
      </a>
      <ul class="nav section-nav flex-column">
       <li class="toc-h5 nav-item toc-entry">
        <a class="reference internal nav-link" href="#Named Arguments_repeat2">
         Named Arguments
        </a>
       </li>
      </ul>
     </li>
     <li class="toc-h4 nav-item toc-entry">
      <a class="reference internal nav-link" href="#umap">
       umap
      </a>
      <ul class="nav section-nav flex-column">
       <li class="toc-h5 nav-item toc-entry">
        <a class="reference internal nav-link" href="#Named Arguments_repeat3">
         Named Arguments
        </a>
       </li>
      </ul>
     </li>
     <li class="toc-h4 nav-item toc-entry">
      <a class="reference internal nav-link" href="#mds">
       mds
      </a>
      <ul class="nav section-nav flex-column">
       <li class="toc-h5 nav-item toc-entry">
        <a class="reference internal nav-link" href="#Named Arguments_repeat4">
         Named Arguments
        </a>
       </li>
      </ul>
     </li>
    </ul>
   </li>
  </ul>
 </li>
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#module-src.embed">
   API
  </a>
 </li>
</ul>

            </nav>
        </div>
    </div>
</div>
    <div id="main-content" class="row">
        <div class="col-12 col-md-9 pl-md-3 pr-md-0">
        
              <div>
                
  <div class="section" id="src-embed-module">
<h1>src.embed module<a class="headerlink" href="#src-embed-module" title="Permalink to this headline">¶</a></h1>
<div class="section" id="command-line-interface">
<h2>Command line interface<a class="headerlink" href="#command-line-interface" title="Permalink to this headline">¶</a></h2>
<p><p>Reduced dimension embeddings for pathogen sequences</p>
</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">usage</span><span class="p">:</span> <span class="n">embed</span> <span class="p">[</span><span class="o">-</span><span class="n">h</span><span class="p">]</span> <span class="p">[</span><span class="o">--</span><span class="n">distance</span><span class="o">-</span><span class="n">matrix</span> <span class="n">DISTANCE_MATRIX</span><span class="p">]</span> <span class="p">[</span><span class="o">--</span><span class="n">separator</span> <span class="n">SEPARATOR</span><span class="p">]</span>
             <span class="p">[</span><span class="o">--</span><span class="n">alignment</span> <span class="n">ALIGNMENT</span><span class="p">]</span> <span class="p">[</span><span class="o">--</span><span class="n">cluster</span><span class="o">-</span><span class="n">data</span> <span class="n">CLUSTER_DATA</span><span class="p">]</span>
             <span class="p">[</span><span class="o">--</span><span class="n">cluster</span><span class="o">-</span><span class="n">threshold</span> <span class="n">CLUSTER_THRESHOLD</span><span class="p">]</span>
             <span class="p">[</span><span class="o">--</span><span class="n">random</span><span class="o">-</span><span class="n">seed</span> <span class="n">RANDOM_SEED</span><span class="p">]</span> <span class="p">[</span><span class="o">--</span><span class="n">output</span><span class="o">-</span><span class="n">dataframe</span> <span class="n">OUTPUT_DATAFRAME</span><span class="p">]</span>
             <span class="p">[</span><span class="o">--</span><span class="n">output</span><span class="o">-</span><span class="n">figure</span> <span class="n">OUTPUT_FIGURE</span><span class="p">]</span>
             <span class="p">{</span><span class="n">pca</span><span class="p">,</span><span class="n">t</span><span class="o">-</span><span class="n">sne</span><span class="p">,</span><span class="n">umap</span><span class="p">,</span><span class="n">mds</span><span class="p">}</span> <span class="o">...</span>
</pre></div>
</div>
<div class="section" id="Positional Arguments">
<h3>Positional Arguments<a class="headerlink" href="#Positional Arguments" title="Permalink to this headline">¶</a></h3>
<dl class="option-list">
<dt><kbd>command</kbd></dt>
<dd><p>Possible choices: pca, t-sne, umap, mds</p>
</dd>
</dl>
</div>
<div class="section" id="Named Arguments">
<h3>Named Arguments<a class="headerlink" href="#Named Arguments" title="Permalink to this headline">¶</a></h3>
<dl class="option-list">
<dt><kbd>--distance-matrix</kbd></dt>
<dd><p>a distance matrix that can be read in by pandas, index column as row 0</p>
</dd>
<dt><kbd>--separator</kbd></dt>
<dd><p>separator between columns in the given distance matrix</p>
<p>Default: “,”</p>
</dd>
<dt><kbd>--alignment</kbd></dt>
<dd><p>an aligned FASTA file to create a distance matrix with. Make sure the strain order in this file matches the order in the distance matrix.</p>
</dd>
<dt><kbd>--cluster-data</kbd></dt>
<dd><p>The file (same separator as distance-matrix) that contains the distance threshold by which to cluster data in the embedding and assign labels given via HDBSCAN (<a class="reference external" href="https://hdbscan.readthedocs.io/en/latest/how_hdbscan_works.html">https://hdbscan.readthedocs.io/en/latest/how_hdbscan_works.html</a>). If no value is given in cluster-data or cluster-threshold, the default distance threshold of 0.0 will be used.</p>
</dd>
<dt><kbd>--cluster-threshold</kbd></dt>
<dd><p>The float value for the distance threshold by which to cluster data in the embedding and assign labels via HDBSCAN. If no value is given in cluster-data or cluster-threshold, the default distance threshold of 0.0 will be used.</p>
</dd>
<dt><kbd>--random-seed</kbd></dt>
<dd><p>an integer used for reproducible results.</p>
<p>Default: 314159</p>
</dd>
<dt><kbd>--output-dataframe</kbd></dt>
<dd><p>a csv file outputting the embedding with the strain name and its components.</p>
</dd>
<dt><kbd>--output-figure</kbd></dt>
<dd><p>outputs a PNG plot of the embedding</p>
</dd>
</dl>
</div>
<div class="section" id="Sub-commands:">
<h3>Sub-commands:<a class="headerlink" href="#Sub-commands:" title="Permalink to this headline">¶</a></h3>
<div class="section" id="pca">
<h4>pca<a class="headerlink" href="#pca" title="Permalink to this headline">¶</a></h4>
<p>Principal Component Analysis</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">embed</span> <span class="n">pca</span> <span class="p">[</span><span class="o">-</span><span class="n">h</span><span class="p">]</span> <span class="p">[</span><span class="o">--</span><span class="n">components</span> <span class="n">COMPONENTS</span><span class="p">]</span>
          <span class="p">[</span><span class="o">--</span><span class="n">explained</span><span class="o">-</span><span class="n">variance</span> <span class="n">EXPLAINED_VARIANCE</span><span class="p">]</span>
</pre></div>
</div>
<div class="section" id="Named Arguments_repeat1">
<h5>Named Arguments<a class="headerlink" href="#Named Arguments_repeat1" title="Permalink to this headline">¶</a></h5>
<dl class="option-list">
<dt><kbd>--components</kbd></dt>
<dd><p>the number of components for PCA</p>
<p>Default: 10</p>
</dd>
<dt><kbd>--explained-variance</kbd></dt>
<dd><p>the path for the CSV explained variance for each component</p>
</dd>
</dl>
</div>
</div>
<div class="section" id="t-sne">
<h4>t-sne<a class="headerlink" href="#t-sne" title="Permalink to this headline">¶</a></h4>
<p>t-distributed Stochastic Neighborhood Embedding</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">embed</span> <span class="n">t</span><span class="o">-</span><span class="n">sne</span> <span class="p">[</span><span class="o">-</span><span class="n">h</span><span class="p">]</span> <span class="p">[</span><span class="o">--</span><span class="n">perplexity</span> <span class="n">PERPLEXITY</span><span class="p">]</span> <span class="p">[</span><span class="o">--</span><span class="n">learning</span><span class="o">-</span><span class="n">rate</span> <span class="n">LEARNING_RATE</span><span class="p">]</span>
</pre></div>
</div>
<div class="section" id="Named Arguments_repeat2">
<h5>Named Arguments<a class="headerlink" href="#Named Arguments_repeat2" title="Permalink to this headline">¶</a></h5>
<dl class="option-list">
<dt><kbd>--perplexity</kbd></dt>
<dd><p>The perplexity is related to the number of nearest neighbors. Because of this, the size of the dataset is proportional to the best perplexity value (large dataset -&gt; large perplexity). Values between 5 and 50 work best. The default value is the value consistently the best for pathogen analyses, results from an exhaustive grid search.</p>
<p>Default: 30.0</p>
</dd>
<dt><kbd>--learning-rate</kbd></dt>
<dd><p>The learning rate for t-SNE is usually between 10.0 and 1000.0. Values out of these bounds may create innacurate results. The default value is the value consistently the best for pathogen analyses, results from an exhaustive grid search.</p>
<p>Default: 200.0</p>
</dd>
</dl>
</div>
</div>
<div class="section" id="umap">
<h4>umap<a class="headerlink" href="#umap" title="Permalink to this headline">¶</a></h4>
<p>Uniform Manifold Approximation and Projection</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">embed</span> <span class="n">umap</span> <span class="p">[</span><span class="o">-</span><span class="n">h</span><span class="p">]</span> <span class="p">[</span><span class="o">--</span><span class="n">nearest</span><span class="o">-</span><span class="n">neighbors</span> <span class="n">NEAREST_NEIGHBORS</span><span class="p">]</span> <span class="p">[</span><span class="o">--</span><span class="nb">min</span><span class="o">-</span><span class="n">dist</span> <span class="n">MIN_DIST</span><span class="p">]</span>
</pre></div>
</div>
<div class="section" id="Named Arguments_repeat3">
<h5>Named Arguments<a class="headerlink" href="#Named Arguments_repeat3" title="Permalink to this headline">¶</a></h5>
<dl class="option-list">
<dt><kbd>--nearest-neighbors</kbd></dt>
<dd><p>Nearest neighbors controls how UMAP balances local versus global structure in the data (finer detail patterns versus global structure). This value is proportional to the size of the data (large dataset -&gt; large nearest neighbors. The default value is the value consistently the best for pathogen analyses, results from an exhaustive grid search.</p>
<p>Default: 200</p>
</dd>
<dt><kbd>--min-dist</kbd></dt>
<dd><p>Minimum Distance controls how tightly packed the UMAP embedding is. While it does not change the structure of the data, it does change the embedding’s shape. The default value is the value consistently the best for pathogen analyses, results from an exhaustive grid search.</p>
<p>Default: 0.5</p>
</dd>
</dl>
</div>
</div>
<div class="section" id="mds">
<h4>mds<a class="headerlink" href="#mds" title="Permalink to this headline">¶</a></h4>
<p>Multidimensional Scaling</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">embed</span> <span class="n">mds</span> <span class="p">[</span><span class="o">-</span><span class="n">h</span><span class="p">]</span> <span class="p">[</span><span class="o">--</span><span class="n">components</span> <span class="n">COMPONENTS</span><span class="p">]</span>
</pre></div>
</div>
<div class="section" id="Named Arguments_repeat4">
<h5>Named Arguments<a class="headerlink" href="#Named Arguments_repeat4" title="Permalink to this headline">¶</a></h5>
<dl class="option-list">
<dt><kbd>--components</kbd></dt>
<dd><p>the number of components for MDS</p>
<p>Default: 10</p>
</dd>
</dl>
</div>
</div>
</div>
</div>
<div class="section" id="module-src.embed">
<span id="api"></span><h2>API<a class="headerlink" href="#module-src.embed" title="Permalink to this headline">¶</a></h2>
<dl class="py function">
<dt id="src.embed.get_hamming_distances">
<code class="sig-prename descclassname"><span class="pre">src.embed.</span></code><code class="sig-name descname"><span class="pre">get_hamming_distances</span></code><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">genomes</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#src.embed.get_hamming_distances" title="Permalink to this definition">¶</a></dt>
<dd><p>Calculate pairwise Hamming distances between the given list of genomes
and return the nonredundant array of values for use with scipy’s squareform function.
Bases other than standard nucleotides (A, T, C, G) are ignored.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>genomes</strong> (<em>list</em>) – a list of strings corresponding to genomes that should be compared</p>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>a list of distinct Hamming distances as a vector-form distance vector</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p>list</p>
</dd>
</dl>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">genomes</span> <span class="o">=</span> <span class="p">[</span><span class="s2">&quot;ATGCT&quot;</span><span class="p">,</span> <span class="s2">&quot;ATGCT&quot;</span><span class="p">,</span> <span class="s2">&quot;ACGCT&quot;</span><span class="p">]</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">get_hamming_distances</span><span class="p">(</span><span class="n">genomes</span><span class="p">)</span>
<span class="go">[0, 1, 1]</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">genomes</span> <span class="o">=</span> <span class="p">[</span><span class="s2">&quot;AT-GCT&quot;</span><span class="p">,</span> <span class="s2">&quot;AT--CT&quot;</span><span class="p">,</span> <span class="s2">&quot;AC--CT&quot;</span><span class="p">]</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">get_hamming_distances</span><span class="p">(</span><span class="n">genomes</span><span class="p">)</span>
<span class="go">[0, 1, 1]</span>
</pre></div>
</div>
</dd></dl>

<dl class="py function">
<dt id="src.embed.make_parser">
<code class="sig-prename descclassname"><span class="pre">src.embed.</span></code><code class="sig-name descname"><span class="pre">make_parser</span></code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#src.embed.make_parser" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

</div>
</div>


              </div>
              
        
            



<div class='prev-next-bottom'>
    

</div>
        
        </div>
    </div>
    <footer class="footer mt-5 mt-md-0">
    <div class="container">
      <p>
        
          By Sravani Nanduri and John Huddleston<br/>
        
            &copy; Copyright 2021–2021 Sravani Nanduri.<br/>
      </p>
    </div>
  </footer>
</main>


      </div>
    </div>
  
  <script src="static/js/index.1c5a1a01449ed65a7b51.js"></script>

  
  </body>
</html>